base/array.jl

```diff
@@ -1492,26 +1492,45 @@ cat(n::Integer, x::Integer...) = reshape([x...], (ntuple(x->1, n-1)..., length(x
 
 ## find ##
 
+_pairs(A::Union{AbstractArray, AbstractDict, AbstractString, Tuple, NamedTuple}) = pairs(A)
+_pairs(iter) = _pairs(IteratorSize(iter), iter)
+_pairs(::Union{HasLength, HasShape}, iter) = zip(1:length(iter), iter)
```

Comment:
When writing this I realized we could return `CartesianIndex` objects for `HasShape` iterators. But for consistency with arrays (and for convenience), we should return `Int` linear indices rather than one-dimensional `CartesianIndex` objects for 1D `HasShape` iterators. That's not possible at the moment without adding dimension type parameter, i.e. `HasShape{N}` as in https://github.com/JuliaLang/julia/pull/25356#discussion_r159224573. Can easily be changed in a subsequent PR.

Reply:
Filed https://github.com/JuliaLang/julia/pull/25655.